crates/net/downloaders/src/bodies/concurrent.rs

```diff
@@ -131,41 +131,39 @@ where
             return Ok(headers.into_iter().cloned().map(BlockResponse::Empty).collect())
         }
 
-        let request_len = headers_with_txs_and_ommers.len();
-        tracing::trace!(target: "downloaders::bodies", request_len, "Requesting bodies");
-        let (peer_id, bodies) =
-            self.client.get_block_bodies(headers_with_txs_and_ommers).await?.split();
-        tracing::trace!(
-            target: "downloaders::bodies", request_len, response_len = bodies.len(), ?peer_id, "Received bodies"
-        );
+        let mut responses = Vec::with_capacity(headers_with_txs_and_ommers.len());
+        while responses.len() != headers_with_txs_and_ommers.len() {
+            let request: Vec<_> =
+                headers_with_txs_and_ommers.iter().skip(responses.len()).cloned().collect();
+            let request_len = request.len();
+            tracing::trace!(target: "downloaders::bodies", request_len, "Requesting bodies");
+            let (peer_id, bodies) = self.client.get_block_bodies(request.clone()).await?.split();
```

Comment:
this seems fine,
I guess you could get rid of the first redundant collect by flipping this to 

```rust
let mut request = headers_with_txs_and_ommers.clone();
loop {

  // request
  
  // break if `responses.len() != headers_with_txs_and_ommers.len()`  else update request

}
```

Reply:
but I guess iter+collect vs clone is negligible

Reply:
nvm the existing solution is better